### ISSUE TITLE
Bash scripts spawned by a pm2-managed node application hang

### ISSUE THREAD

**Issue by joaquimds** [a2o/snoopy#117]
## Checklist before I submit this issue report

I confirm that:
- [x] I have tested this with the latest version of Snoopy available @ https://github.com/a2o/snoopy#i1-version-information-and-download-links (not the latest Snoopy version packaged by your favorite distribution, it might be out of date)
- [x] I have read Snoopy documentation @ https://github.com/a2o/snoopy/blob/master/README.md & https://github.com/a2o/snoopy/tree/master/doc
- [x] I have checked FAQ entries @ https://github.com/a2o/snoopy/blob/master/doc/FAQ.md
- [x] I have checked existing Snoopy issues for duplicates and found none @ https://github.com/a2o/snoopy/issues?q=is%3Aissue

## Issue description - Environment basics

| Key                                 | Value             | Comments                         |
| ----------------------------------- | ----------------- | ---------------------------------|
| Architecture:                       | x86_64            | x86 is not supported anymore     |
| Linux distribution:                 | Ubuntu            | |
| Distribution version:               | 14.04            | |
| Snoopy version:                     | 2.4.6            | |
| Snoopy config file was used:        | yes   | snoopy-install.sh enables this by default |
| Snoopy threading support enabled:   | no   | not enabled by default as of v2.4.4       |

## Issue description - How was Snoopy installed

snoopy-install.sh

## Issue description - Describe the problem you are having

We use pm2 (https://github.com/Unitech/pm2) to keep our node processes running. We noticed that pm2 spawns a lot of processes that do a lot of logging, so we have added `exclude_spawns_of:pm2;` to our snoopy.ini file. Unfortunately this seems to break bash scripts that are spawned by our node application. The bash scripts start up, but can't run any commands – all that happens is a child bash process is created that quickly takes up 100% of the machine's CPU and never terminates.

## Steps to reproduce
1. Install snoopy and pm2
2. Modify snoopy.ini to include `exclude_spawns_of:pm2;`
3. Put `test.js` and `test.sh` in a folder.
4. In this folder, run `pm2 start test.js`
5. Observe that the bash process started by `test.js` hangs forever.

#### test.js
    'use strict';

    const child_process = require('child_process');

    const process = child_process.spawn('./test.sh');
    process.on('exit', () => {
        console.log('process exit');
    });
    process.stdout.on('data', (data) => {
        console.log('Output: ' + data.toString('utf8'));
    });

#### test.sh
    #!/usr/bin/env bash

    TEMP=$(mktemp -d);
    echo "Created directory $TEMP"
    rm -rf ${TEMP}
    echo "Deleted directory $TEMP"

**Comment by akwizgran** [comment]
I've found the problem. The name of the PM2 process is "PM2 v2.3.0: God", including whitespace. PM2 adds the working directory to the name of the Node process it spawns, so that process has a name like "node /foo", again including whitespace.

The whitespace causes a matching failure for the fscanf call at line 145 of exclude_spawns_of.c. The return value isn't checked for matching failures, only for EOF.

In the case of the PM2 process, this causes the PPID to be wrongly interpreted as 2. On my test system, PID 2 belongs to kthreadd, which has PPID 0. When the filter fails to open /proc/0/stat, it returns. So for children of the PM2 process, the filter examines the wrong chain of ancestor processes but it still terminates.

However, in the case of the Node process spawned by PM2, the matching failure causes fscanf not to update the PPID. This causes an infinite loop when a descendent of the Node process spawns a child.

**Comment by bostjan** [comment]
@akwizgran  Thanks for the report and pull request. Please run the test suite locally and update here with the results (or in pull request comments).

**Comment by akwizgran** [comment]
@bostjan here's the result of running the test suite:
```
$ make test
Making check in etc
make[1]: Entering directory '/home/someone/snoopy/etc'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/home/someone/snoopy/etc'
Making check in lib
make[1]: Entering directory '/home/someone/snoopy/lib'
Making check in inih
make[2]: Entering directory '/home/someone/snoopy/lib/inih'
Making check in src
make[3]: Entering directory '/home/someone/snoopy/lib/inih/src'
make[3]: Nothing to be done for 'check'.
make[3]: Leaving directory '/home/someone/snoopy/lib/inih/src'
make[3]: Entering directory '/home/someone/snoopy/lib/inih'
make[3]: Nothing to be done for 'check-am'.
make[3]: Leaving directory '/home/someone/snoopy/lib/inih'
make[2]: Leaving directory '/home/someone/snoopy/lib/inih'
make[2]: Entering directory '/home/someone/snoopy/lib'
make[2]: Nothing to be done for 'check-am'.
make[2]: Leaving directory '/home/someone/snoopy/lib'
make[1]: Leaving directory '/home/someone/snoopy/lib'
Making check in src
make[1]: Entering directory '/home/someone/snoopy/src'
Making check in eventsource
make[2]: Entering directory '/home/someone/snoopy/src/eventsource'
make[2]: Nothing to be done for 'check'.
make[2]: Leaving directory '/home/someone/snoopy/src/eventsource'
Making check in datasource
make[2]: Entering directory '/home/someone/snoopy/src/datasource'
make[2]: Nothing to be done for 'check'.
make[2]: Leaving directory '/home/someone/snoopy/src/datasource'
Making check in filter
make[2]: Entering directory '/home/someone/snoopy/src/filter'
make[2]: Nothing to be done for 'check'.
make[2]: Leaving directory '/home/someone/snoopy/src/filter'
Making check in output
make[2]: Entering directory '/home/someone/snoopy/src/output'
make[2]: Nothing to be done for 'check'.
make[2]: Leaving directory '/home/someone/snoopy/src/output'
Making check in .
make[2]: Entering directory '/home/someone/snoopy/src'
make[2]: Nothing to be done for 'check-am'.
make[2]: Leaving directory '/home/someone/snoopy/src'
make[1]: Leaving directory '/home/someone/snoopy/src'
Making check in util
make[1]: Entering directory '/home/someone/snoopy/util'
make[1]: Nothing to be done for 'check'.
make[1]: Leaving directory '/home/someone/snoopy/util'
Making check in doc
make[1]: Entering directory '/home/someone/snoopy/doc'
Making check in internals
make[2]: Entering directory '/home/someone/snoopy/doc/internals'
make[2]: Nothing to be done for 'check'.
make[2]: Leaving directory '/home/someone/snoopy/doc/internals'
make[2]: Entering directory '/home/someone/snoopy/doc'
make[2]: Nothing to be done for 'check-am'.
make[2]: Leaving directory '/home/someone/snoopy/doc'
make[1]: Leaving directory '/home/someone/snoopy/doc'
Making check in tests
make[1]: Entering directory '/home/someone/snoopy/tests'
Making check in bin
make[2]: Entering directory '/home/someone/snoopy/tests/bin'
make[2]: Nothing to be done for 'check'.
make[2]: Leaving directory '/home/someone/snoopy/tests/bin'
Making check in configfile
make[2]: Entering directory '/home/someone/snoopy/tests/configfile'
make  check-TESTS
make[3]: Entering directory '/home/someone/snoopy/tests/configfile'
make[4]: Entering directory '/home/someone/snoopy/tests/configfile'
PASS: configfile-filter_chain-1.sh
PASS: configfile-message_format-1.sh
PASS: configfile-output-devlog.sh
PASS: configfile-output-file.sh
XFAIL: configfile-output-invalid.sh
PASS: configfile-output-socket.sh
PASS: configfile-syslog_facility.sh
PASS: configfile-syslog_ident.sh
PASS: configfile-syslog_level.sh
make[5]: Entering directory '/home/someone/snoopy/tests/configfile'
make[5]: Nothing to be done for 'all'.
make[5]: Leaving directory '/home/someone/snoopy/tests/configfile'
============================================================================
Testsuite summary for Snoopy Logger 2.4.6-2-g877cb35
============================================================================
# TOTAL: 9
# PASS:  8
# SKIP:  0
# XFAIL: 1
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[4]: Leaving directory '/home/someone/snoopy/tests/configfile'
make[3]: Leaving directory '/home/someone/snoopy/tests/configfile'
make[2]: Leaving directory '/home/someone/snoopy/tests/configfile'
Making check in datasource
make[2]: Entering directory '/home/someone/snoopy/tests/datasource'
make  check-TESTS
make[3]: Entering directory '/home/someone/snoopy/tests/datasource'
make[4]: Entering directory '/home/someone/snoopy/tests/datasource'
PASS: datasource_cmdline.sh
PASS: datasource_cwd.sh
PASS: datasource_datetime.sh
PASS: datasource_egid.sh
PASS: datasource_egroup.sh
PASS: datasource_env_all.sh
PASS: datasource_env-set.sh
PASS: datasource_env-unset.sh
PASS: datasource_euid.sh
PASS: datasource_eusername.sh
PASS: datasource_filename.sh
PASS: datasource_gid.sh
PASS: datasource_group.sh
PASS: datasource_hostname.sh
PASS: datasource_login.sh
PASS: datasource_pid.sh
PASS: datasource_ppid.sh
PASS: datasource_rpname.sh
PASS: datasource_sid.sh
PASS: datasource_snoopy_literal.sh
PASS: datasource_snoopy_version.sh
PASS: datasource_tid.sh
PASS: datasource_tid_kernel.sh
PASS: datasource_timestamp.sh
PASS: datasource_timestamp_ms.sh
PASS: datasource_timestamp_us.sh
PASS: datasource_tty.sh
PASS: datasource_tty_uid.sh
PASS: datasource_tty_username.sh
PASS: datasource_uid.sh
PASS: datasource_username.sh
make[5]: Entering directory '/home/someone/snoopy/tests/datasource'
make[5]: Nothing to be done for 'all'.
make[5]: Leaving directory '/home/someone/snoopy/tests/datasource'
============================================================================
Testsuite summary for Snoopy Logger 2.4.6-2-g877cb35
============================================================================
# TOTAL: 31
# PASS:  31
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[4]: Leaving directory '/home/someone/snoopy/tests/datasource'
make[3]: Leaving directory '/home/someone/snoopy/tests/datasource'
make[2]: Leaving directory '/home/someone/snoopy/tests/datasource'
Making check in filter
make[2]: Entering directory '/home/someone/snoopy/tests/filter'
make  check-TESTS
make[3]: Entering directory '/home/someone/snoopy/tests/filter'
make[4]: Entering directory '/home/someone/snoopy/tests/filter'
PASS: filter_exclude_uid-pass.sh
PASS: filter_exclude_uid-pass-multi.sh
PASS: filter_exclude_uid-drop.sh
PASS: filter_exclude_uid-drop-multi.sh
PASS: filter_only_tty.sh
PASS: filter_only_uid-pass.sh
PASS: filter_only_uid-pass-multi.sh
PASS: filter_only_uid-drop.sh
PASS: filter_only_uid-drop-multi.sh
make[5]: Entering directory '/home/someone/snoopy/tests/filter'
make[5]: Nothing to be done for 'all'.
make[5]: Leaving directory '/home/someone/snoopy/tests/filter'
============================================================================
Testsuite summary for Snoopy Logger 2.4.6-2-g877cb35
============================================================================
# TOTAL: 9
# PASS:  9
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[4]: Leaving directory '/home/someone/snoopy/tests/filter'
make[3]: Leaving directory '/home/someone/snoopy/tests/filter'
make[2]: Leaving directory '/home/someone/snoopy/tests/filter'
Making check in message
make[2]: Entering directory '/home/someone/snoopy/tests/message'
make  check-TESTS
make[3]: Entering directory '/home/someone/snoopy/tests/message'
make[4]: Entering directory '/home/someone/snoopy/tests/message'
PASS: message_format-datasource-arg.sh
PASS: message_format-datasource-noarg.sh
PASS: message_format-empty.sh
PASS: message_format-static.sh
make[5]: Entering directory '/home/someone/snoopy/tests/message'
make[5]: Nothing to be done for 'all'.
make[5]: Leaving directory '/home/someone/snoopy/tests/message'
============================================================================
Testsuite summary for Snoopy Logger 2.4.6-2-g877cb35
============================================================================
# TOTAL: 4
# PASS:  4
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[4]: Leaving directory '/home/someone/snoopy/tests/message'
make[3]: Leaving directory '/home/someone/snoopy/tests/message'
make[2]: Leaving directory '/home/someone/snoopy/tests/message'
Making check in output
make[2]: Entering directory '/home/someone/snoopy/tests/output'
make  check-TESTS
make[3]: Entering directory '/home/someone/snoopy/tests/output'
make[4]: Entering directory '/home/someone/snoopy/tests/output'
PASS: output_devnull.sh
PASS: output_file.sh
PASS: output_socket.sh
PASS: output_stderr.sh
PASS: output_stdout.sh
make[5]: Entering directory '/home/someone/snoopy/tests/output'
make[5]: Nothing to be done for 'all'.
make[5]: Leaving directory '/home/someone/snoopy/tests/output'
============================================================================
Testsuite summary for Snoopy Logger 2.4.6-2-g877cb35
============================================================================
# TOTAL: 5
# PASS:  5
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[4]: Leaving directory '/home/someone/snoopy/tests/output'
make[3]: Leaving directory '/home/someone/snoopy/tests/output'
make[2]: Leaving directory '/home/someone/snoopy/tests/output'
Making check in combined
make[2]: Entering directory '/home/someone/snoopy/tests/combined'
make  check-TESTS
make[3]: Entering directory '/home/someone/snoopy/tests/combined'
make[4]: Entering directory '/home/someone/snoopy/tests/combined'
PASS: combined-nofilter-1.sh
PASS: combined-nofilter-2.sh
PASS: combined-filter-exclude_uid-drop.sh
PASS: combined-filter-exclude_uid-pass.sh
PASS: combined-filter-only_uid-drop.sh
PASS: combined-filter-only_uid-pass.sh
make[5]: Entering directory '/home/someone/snoopy/tests/combined'
make[5]: Nothing to be done for 'all'.
make[5]: Leaving directory '/home/someone/snoopy/tests/combined'
============================================================================
Testsuite summary for Snoopy Logger 2.4.6-2-g877cb35
============================================================================
# TOTAL: 6
# PASS:  6
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[4]: Leaving directory '/home/someone/snoopy/tests/combined'
make[3]: Leaving directory '/home/someone/snoopy/tests/combined'
make[2]: Leaving directory '/home/someone/snoopy/tests/combined'
make[2]: Entering directory '/home/someone/snoopy/tests'
make[2]: Nothing to be done for 'check-am'.
make[2]: Leaving directory '/home/someone/snoopy/tests'
make[1]: Leaving directory '/home/someone/snoopy/tests'
make[1]: Entering directory '/home/someone/snoopy'
make[1]: Leaving directory '/home/someone/snoopy'
```

**Comment by bostjan** [comment]
Excellent.
I'll review your PR #118 and merge it for the next release.
(It is a bit bigger update to the code, that is why I didn't merge it right away.)

**Comment by bostjan** [comment]
Sorry it took me a while to get to this, but now this is already merged and will land in 2.4.10.

Thanks again @akwizgran for the PR and for the excellent description of the problem, much appreciated.
Thanks @joaquimds for reporting this.
